src/libraries/Microsoft.PowerFx.LanguageServerProtocol/ScopeVariations/IAsyncPowerFxScope.cs

```diff
@@ -0,0 +1,40 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.PowerFx.Intellisense;
+
+namespace Microsoft.PowerFx.LanguageServerProtocol
+{
+    /// <summary>
+    /// Asynchronous version of <see cref="IPowerFxScope"/>.
+    /// This allows consumers of SDK to asynchronously perform scope operations.
+    /// </summary>
+    public interface IAsyncPowerFxScope : IPowerFxScope
+    {
+        /// <summary>
+        /// Check for errors in the given expression. 
+        /// </summary>
+        /// <param name="expression">The expression to validate.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        /// <returns>CheckResult.</returns>
+        Task<CheckResult> CheckAsync(string expression, CancellationToken cancellationToken);

```

Comment:
>CheckAsync [](http://example.com/codeflow?start=26&length=10)

Why would Check be async?  This normally should be an entirely sync method.


I could imagine SuggestAsync being async if we support dynamic intellisense.

Reply:
Yes computing check is synchronous ideally but in Document Server, we do this behind a lock and putting it behind a lock in document server makes the check result computation effectively asynchronous. A host can have requirements like this where they wrap traditional check result computation in a async code path.
An example:
![image](https://github.com/microsoft/Power-Fx/assets/25670945/92d7291d-92cd-47eb-b9f9-12f00103820b)